src/core/image_resizer.js

```diff
@@ -0,0 +1,326 @@
+/* Copyright 2023 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { FeatureTest, ImageKind, shadow } from "../shared/util.js";
+
+const MIN_IMAGE_DIM = 2048;
+const MAX_IMAGE_DIM = 65536;
+
+// Large images are encoded in using the BMP format (it's a way faster than
+// encoding in PNG because there are no checksums to compute).
+// Unfortunately, createImageBitmap will create a task in the main thread in
+// order to build the bitmap, hence the main thread is blocked during the
+// decoding which can be a bit long (see bug 1817244).
+// When bug 1759728 is fixed we can remove the BMP workaround and at least it
+// should be a way faster to create the bitmap.
+
+class ImageResizer {
+  constructor(imgData, isMask) {
+    this._imgData = imgData;
+    this._isMask = isMask;
+  }
+
+  static needsToBeResized(width, height) {
+    if (width <= MIN_IMAGE_DIM && height <= MIN_IMAGE_DIM) {
+      return false;
+    }
+    const { MAX_AREA, MAX_DIM } = this;
+    return (
+      Math.max(width / MAX_DIM, height / MAX_DIM, (width * height) / MAX_AREA) >
+      1
+    );
+  }
+
+  static get MAX_DIM() {
+    return shadow(
+      this,
+      "MAX_DIM",
+      this._guessMax(MIN_IMAGE_DIM, MAX_IMAGE_DIM, 1)
+    );
+  }
+
+  static get MAX_AREA() {
+    // TODO: the computation can be a bit long because we potentially allocate
+    // some large canvas, so in the Firefox case this value (and MAX_DIM) can be
+    // infered from prefs (MAX_AREA = gfx.max-alloc-size / 4, 4 is because of
+    // RGBA).
+    // It really must be improved for chrome where it takes ~5s !!
+    return shadow(
+      this,
+      "MAX_AREA",
+      this._guessMax(MIN_IMAGE_DIM, this.MAX_DIM, 0) ** 2
+    );
+  }
+
+  static _guessMax(start, end, defaultHeight) {
+    while (start < end) {
+      const middle = Math.floor((start + end) / 2);
+      const height = defaultHeight || middle;
+      let width;
```

Comment:
Nit: By initializing this to zero the `catch`-handler further down can just be empty.
```suggestion
      let width = 0;
```